File: app/src/main/java/com/camtech/books/utils/ConnectivityReceiver.java
package com.camtech.books.utils;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;

/**
 * Class to listen for changes in internet connection.
 * Since it's not registered in the manifest, it only
 * lives within the instance of the class it's used in.
 * */
public class ConnectivityReceiver extends BroadcastReceiver{

    private Context context;
   private ConnectionListener connectionListener;

    public ConnectivityReceiver(Context context) {

        this.context = context;
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        // Mobile connection has dropped

        if (intent.getAction() != null && intent.getAction().equals(ConnectivityManager.CONNECTIVITY_ACTION)) {
            if (!hasConnection()) {
                if (connectionListener != null)connectionListener.onConnectionDropped();
            } else {
                // Connection has been established
               if (connectionListener != null) connectionListener.onConnect();
            }
        }
    }

    public boolean hasConnection(){
        ConnectivityManager manager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo networkInfo = null;
        if (manager != null) {
            networkInfo = manager.getActiveNetworkInfo();
        }

        return networkInfo != null && networkInfo.isConnectedOrConnecting();
    }

    public void setConnectionListener(ConnectionListener connectionListener) {
        this.connectionListener = connectionListener;
    }
}

File: app/src/main/java/com/camtech/books/utils/ConnectivityReceiver.kt
package com.camtech.books.utils

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager

/**
 * Class to listen for changes in internet connection.
 * Since it's not registered in the manifest, it only
 * lives within the instance of the class it's used in.
 */
class ConnectivityReceiver(private val context: Context) : BroadcastReceiver() {
    private var connectionListener: ConnectionListener? = null

    override fun onReceive(context: Context?, intent: Intent?) {
        if (intent?.action == ConnectivityManager.CONNECTIVITY_ACTION) {
            if (!hasConnection()) connectionListener?.onConnectionDropped()
            else connectionListener?.onConnect()
        }
    }

    fun hasConnection(): Boolean {
        val manager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val networkInfo = manager.activeNetworkInfo
        return networkInfo != null && networkInfo.isConnectedOrConnecting
    }

    fun setConnectionListener(connectionListener: ConnectionListener) {
        this.connectionListener = connectionListener
    }
}